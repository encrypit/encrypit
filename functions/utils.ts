import { EXPIRATION } from 'shared/constants';
import type { CustomMetadata } from 'shared/types';

import type { Context, Env } from './types';

/**
 * Gets R2 bucket.
 *
 * @param context - Event context.
 * @param expirationDays - Expiration days.
 * @returns - R2 bucket.
 */
export function getBucket(
  context: Context,
  expirationDays = EXPIRATION.DAYS_7
): R2Bucket {
  switch (expirationDays) {
    case EXPIRATION.DAYS_7:
    default:
      return context.env.EXPIRATION_DAYS_7;
  }
}

/**
 * Gets object custom metadata.
 *
 * @param context - R2 object.
 * @returns - Custom metadata.
 */
export function getCustomMetadata(obj: R2ObjectBody) {
  return obj.customMetadata as unknown as CustomMetadata;
}

/**
 * Generates response init.
 *
 * @param environment - Node environment.
 * @returns - Response init.
 */
export function getResponseInit(environment: Env['NODE_ENV']) {
  const responseInit: ResponseInit = {};
  if (environment === 'development') {
    responseInit.headers = {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': '*',
      'Access-Control-Allow-Methods': '*',
    };
  }
  return responseInit;
}
