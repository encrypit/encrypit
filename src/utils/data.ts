enum Events {
  error = 'error',
  loadend = 'loadend',
}

/**
 * Converts Blob to Base64.
 *
 * @param blob - Blob.
 * @returns - Base64.
 */
export async function blobToBase64(blob: Blob): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    function loadend() {
      resolve(reader.result as string);
      reader.removeEventListener(Events.loadend, loadend);
    }
    reader.addEventListener(Events.loadend, loadend);

    /* istanbul ignore next */
    function error() {
      reject(reader.error);
      reader.removeEventListener(Events.error, error);
    }
    reader.addEventListener(Events.error, error);

    reader.readAsDataURL(blob);
  });
}
