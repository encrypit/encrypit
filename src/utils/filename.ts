/**
 * Generates file name with extension.
 *
 * @returns - File name.
 */
export function generateFileName(): string {
  return `encrypit-download-${getISOString()}.zip`;
}

/**
 * Gets clean ISO string.
 *
 * @returns - ISO with special characters removed.
 */
function getISOString(): string {
  const date = new Date();

  return date
    .toISOString()
    .split('.')[0]
    .replaceAll('-', '')
    .replaceAll(':', '');
}
