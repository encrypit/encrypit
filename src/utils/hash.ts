import { digest } from 'pepto';

const algorithm = 'SHA-512';

/**
 * Hashes password into SHA-512.
 *
 * @param password - Password.
 * @returns - SHA-512.
 */
export function hashPassword(password: string): Promise<string> {
  return digest(algorithm, password);
}
