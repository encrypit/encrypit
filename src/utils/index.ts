export * from './filename';
export * from './form';
export * from './hash';
export * from './zip';
