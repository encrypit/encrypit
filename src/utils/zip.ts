import { BlobReader, BlobWriter, ZipReader, ZipWriter } from '@zip.js/zip.js';
import { MIME } from 'shared/constants';

const zipWriterOptions = {
  zip64: true,
  level: 9,
  bufferedWrite: true,
};

/**
 * Creates zip file.
 *
 * @param files - Files.
 * @param password - Password.
 * @returns - Zip file.
 */
async function zip(files: File[], password?: string): Promise<Blob> {
  const zipWriter = new ZipWriter(new BlobWriter(MIME.ZIP), {
    ...zipWriterOptions,
    password,
  });
  await Promise.all(
    files.map((file) => zipWriter.add(file.name, new BlobReader(file)))
  );
  return zipWriter.close();
}

/**
 * Creates encrypted zip file.
 *
 * @param files - Files.
 * @param password - Password.
 * @returns - Zip file.
 */
export async function createZipFile(
  files: File[],
  password: string
): Promise<Blob> {
  const blob = await zip(files);
  const file = new File([blob], 'Archive.zip', { type: MIME.ZIP });
  return zip([file], password);
}

/**
 * Unzips encrypted zip file.
 *
 * @param file - Zip file.
 * @param password - Password.
 * @returns - Zip file.
 */
export async function unzip(file: Blob, password: string): Promise<Blob> {
  const blobReader = new BlobReader(file);
  const blobWriter = new BlobWriter();
  const zipReader = new ZipReader(blobReader, { password });
  const [firstFile] = await zipReader.getEntries();
  const data = await firstFile.getData!(blobWriter);
  await zipReader.close();
  return data;
}
