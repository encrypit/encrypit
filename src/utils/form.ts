import { FORM_DATA } from 'shared/constants';
import { APP_VERSION } from 'src/config';

interface Data {
  file: Blob;
  version?: string;
}

/**
 * Creates FormData.
 *
 * @param data - Data.
 * @returns - Form data.
 */
export function createFormData({
  file,
  version = APP_VERSION,
}: Data): FormData {
  const formData = new FormData();
  formData.append(FORM_DATA.FILE, file);
  formData.append(FORM_DATA.VERSION, version);
  return formData;
}
