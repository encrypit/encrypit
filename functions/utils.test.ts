/// <reference types="@cloudflare/workers-types" />
/// <reference types="@types/jest" />
import { getResponseInit } from './utils';

describe('getResponseInit', () => {
  it.each(['development', 'preview', 'production'] as const)(
    'returns response object for environment %p',
    (environment) => {
      expect(getResponseInit(environment)).toMatchSnapshot();
    },
  );
});
