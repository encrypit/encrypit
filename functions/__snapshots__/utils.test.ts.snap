// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getResponseInit returns response object for environment "development" 1`] = `
{
  "headers": {
    "Access-Control-Allow-Headers": "*",
    "Access-Control-Allow-Methods": "*",
    "Access-Control-Allow-Origin": "*",
  },
}
`;

exports[`getResponseInit returns response object for environment "preview" 1`] = `
{
  "headers": {},
}
`;

exports[`getResponseInit returns response object for environment "production" 1`] = `
{
  "headers": {},
}
`;
